### PR TITLE
odo list binding

### LINKED ISSUE #5635: user should be able to list existing service bindings
## User Story
- As an odo user
- I want to be able to list Service Bindings
- So that I know to what services my applications are connected and what binding information is available (env variables or files)

## Acceptance Criteria
- [ ] `odo list binding` should display list of  all service bindings in the current namespace and those in the current devfile.yaml (if command is executed in the component directory) (make sure that the logic can later be reused in `odo list` command)
- [ ] the service bindings that are in the current devfile.yaml should be marked with `*` similarly how the current components are marked
- [ ] `SERVICES` field can contain multiple services (as service binding can bind multiple services to one app
- [ ] `RUNNING IN` field indicates if the service biding is present on the cluster or not, and in what mode `None` means that it doesn't exist on the cluster, this is possible only for binding that are read from the devfile.yaml (they are marked with `*`). `Dev`/`Deploy` indicates if the binding was created using `odo dev` or `odo deploy` command. 
- [ ] json output
- [ ] needs to be able to handle both `binding.operators.coreos.com/v1alpha1` and `servicebinding.io/v1alpha3` ServiceBindings (explore if it is possible to convert `servicebinding.io` to `binding.operators.coreos.com`)

## Example

```
$ odo list bindings

ServiceBindings in the "mynamespace" namespace:
   NAME             Application               SERVICES                                           RUNNING IN  
   backend-mongodb  backend (Deploment)       mymongodb (PrconaServerMongoDB.psmdb.percona.com)  Dev         
*  frontend-redis   frontend (Deployment)     myredis (Redis.redis.redis.opstreelabs.in)         None        
   otherbinding     application (Deployment)  myredis (Redis.redis.redis.opstreelabs.in)         Dev     
```

```
$ odo list bindings  -o json
{
  bindingsInDevfile: ["frontend-mysql"],
  bindings: [ 
  {
    application: {
      version: v1
      group: apps
      kind: Deployment
      name: online-banking
     },
   services: [ 
    {
    version: v1alpha1 
    group: example.com 
    kind: AccountService 
    name: prod-account-service 
    }]
  }
  ]
}
```

/kind user-story

### PR DESCRIPTION
**Pull request by feloy**
**What type of PR is this:**

/kind feature

**What does this PR do / why we need it:**

Add `odo list binding` command

- [x] with human readable output
- [x] with JSON output

**Which issue(s) this PR fixes:**

Fixes #5635 

**PR acceptance criteria:**

- [x] Unit test 

- [x] Integration test 

- [x] Documentation 

**How to test changes / Special notes to the reviewer:**

### CODE REVIEW

File: pkg/binding/list.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
+package binding
+
+import (
+	"github.com/devfile/library/pkg/devfile/parser"
+	"github.com/redhat-developer/odo/pkg/api"
+	"github.com/redhat-developer/odo/pkg/labels"
+)
+
+func (o *BindingClient) ListAllBindings(devfileObj parser.DevfileObj, context string) ([]api.ServiceBinding, []string, error) {
```
Comment (rm3l) [inline pkg/binding/list.go:9]: I think unit-testing this method, if possible, could be helpful. WDYT?

**Review by rm3l — CHANGES_REQUESTED** [review]

File: pkg/binding/list.go (hunk 1 above)
Comment (feloy) [reply pkg/binding/list.go:9]: Yes, you are right, I will see how to add unit tests for this function

File: pkg/api/binding.go [hunk 2]
```diff
@@ -14,23 +17,25 @@ type ServiceBinding struct {
...
-	BindingEnvVars []string `json:"bindingEnvVars,omitempty"`
+	BindingFiles   []string        `json:"bindingFiles,omitempty"`
+	BindingEnvVars []string        `json:"bindingEnvVars,omitempty"`
+	RunningIn      RunningModeList `json:"runningIn,omitempty"`
 }
 
 // ServiceBindingFromBinding returns a common api.ServiceBinding structure
 // from a ServiceBinding.binding.operators.coreos.com/v1alpha1
-func ServiceBindingFromBinding(binding bindingApi.ServiceBinding) ServiceBinding {
+func ServiceBindingFromBinding(client kclient.ClientInterface, binding bindingApi.ServiceBinding) (ServiceBinding, error) {
```
[the remark this answers is not in the document]
Comment (feloy) [reply pkg/api/binding.go]: You are right. I'm moving these functions to kclient package, to keep the api package as simple as possible

File: pkg/binding/list_test.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+	type args struct {
+		devfileObj parser.DevfileObj
+		context    string
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		args    args
+		want    []api.ServiceBinding
+		want1   []string
```
Comment (rm3l) [inline pkg/binding/list_test.go:101]: Surely nitpicking, but could you rename this field name into something more meaningful? I had to look at the implementation in `BindingClient#ListAllBindings` to understand what was expected there..
Same thing in L231 with `got1`.

**Review by rm3l — COMMENTED** [review]
LGTM feature-wise. I added a minor comment in the `list_test.go` package.

/approve

**Review by valaparthvi — CHANGES_REQUESTED** [review]

**Comment by feloy** [comment]
Thanks @valaparthvi for you review

File: tests/integration/cmd_namespace_test.go [hunk 4]
```diff
@@ -180,8 +180,10 @@ ComponentSettings:
 
 		Describe("list "+commandName, func() {
 			It(fmt.Sprintf("should successfully list all the %ss", commandName), func() {
-				out := helper.Cmd("odo", "list", commandName).ShouldPass().Out()
-				Expect(out).To(ContainSubstring(commonVar.Project))
+				Eventually(func() string {
+					out := helper.Cmd("odo", "list", commandName).ShouldPass().Out()
+					return out
+				}, 10*time.Second, 1*time.Second).Should(ContainSubstring(commonVar.Project))
```
Comment (rm3l) [inline tests/integration/cmd_namespace_test.go:186]: This change pertains to #5837 and could be removed from this PR..

**Comment by feloy** [comment]
Unit tests fail with

```
WARNING: DATA RACE
Write at 0x00c000103a21 by goroutine 67:
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1196 +0x21a

Previous write at 0x00c000103a21 by goroutine 68:
  testing.(*common).FailNow()
      /usr/local/go/src/testing/testing.go:740 +0x4f
  testing.(*common).Fatalf()
      /usr/local/go/src/testing/testing.go:815 +0x9d
  testing.(*T).Fatalf()
      <autogenerated>:1 +0x75
  github.com/golang/mock/gomock.(*Controller).Call.func1()
      /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/vendor/github.com/golang/mock/gomock/controller.go:231 +0x731
  github.com/golang/mock/gomock.(*Controller).Call()
      /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/vendor/github.com/golang/mock/gomock/controller.go:247 +0xdc
  github.com/redhat-developer/odo/pkg/kclient.(*MockClientInterface).ExecCMDInContainer()
      /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/pkg/kclient/mock_Client.go:315 +0x369
  github.com/redhat-developer/odo/pkg/remotecmd.ExecuteCommand()
      /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/pkg/remotecmd/remotecmd.go:36 +0x701
  github.com/redhat-developer/odo/pkg/remotecmd.(*kubeExecProcessHandler).StartProcessForCommand.func1()
      /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/pkg/remotecmd/kubeexec.go:101 +0xdb

Goroutine 67 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1238 +0x5d7
  github.com/redhat-developer/odo/pkg/remotecmd.TestKubeExecProcessHandler_StartProcessForCommand()
      /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/pkg/remotecmd/kubexec_test.go:196 +0x6bc
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1193 +0x202

Goroutine 68 (finished) created at:
  github.com/redhat-developer/odo/pkg/remotecmd.(*kubeExecProcessHandler).StartProcessForCommand()
      /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/pkg/remotecmd/kubeexec.go:96 +0x73c
  github.com/redhat-developer/odo/pkg/remotecmd.TestKubeExecProcessHandler_StartProcessForCommand.func3()
      /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/pkg/remotecmd/kubexec_test.go:208 +0x3f1
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1193 +0x202
==================
--- FAIL: TestKubeExecProcessHandler_StartProcessForCommand (10.00s)
    --- FAIL: TestKubeExecProcessHandler_StartProcessForCommand/command_with_all_fields_returned_no_error (10.00s)
        remotecmd.go:36: Unexpected call to *kclient.MockClientInterface.ExecCMDInContainer([my-container my-pod [/bin/sh -c echo $$ > /opt/odo/.odo_devfile_cmd_my-exec-cmd.pid && cd /path/to/working/dir && export ENV_VAR2='value2' ENV_VAR1='value1' && (tail -f /path/to/a/file) 1>>/proc/1/fd/1 2>>/proc/1/fd/2] 0xc00047a030 0xc00047a040 <nil> false]) at /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/pkg/remotecmd/remotecmd.go:36 because: 
            expected call at /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/pkg/remotecmd/kubexec_test.go:185 doesn't match the argument at index 2.
            Got: [/bin/sh -c echo $$ > /opt/odo/.odo_devfile_cmd_my-exec-cmd.pid && cd /path/to/working/dir && export ENV_VAR2='value2' ENV_VAR1='value1' && (tail -f /path/to/a/file) 1>>/proc/1/fd/1 2>>/proc/1/fd/2]
            Want: is equal to [/bin/sh -c echo $$ > /opt/odo/.odo_devfile_cmd_my-exec-cmd.pid && cd /path/to/working/dir && export ENV_VAR1='value1' ENV_VAR2='value2' && (tail -f /path/to/a/file) 1>>/proc/1/fd/1 2>>/proc/1/fd/2]
        kubexec_test.go:219: timeout waiting for output handler to get called
        controller.go:137: missing call(s) to *kclient.MockClientInterface.ExecCMDInContainer(is equal to my-container, is equal to my-pod, is equal to [/bin/sh -c echo $$ > /opt/odo/.odo_devfile_cmd_my-exec-cmd.pid && cd /path/to/working/dir && export ENV_VAR1='value1' ENV_VAR2='value2' && (tail -f /path/to/a/file) 1>>/proc/1/fd/1 2>>/proc/1/fd/2], is anything, is anything, is anything, is anything) /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/pkg/remotecmd/kubexec_test.go:185
        controller.go:137: aborting test due to missing call(s)
        testing.go:1092: race detected during execution of test
    testing.go:1092: race detected during execution of test
FAIL
FAIL	github.com/redhat-developer/odo/pkg/remotecmd	16.453s
```

**Comment by rm3l** [comment]
> Unit tests fail with
> 
> ```
> WARNING: DATA RACE
> Write at 0x00c000103a21 by goroutine 67:
>   testing.tRunner()
>       /usr/local/go/src/testing/testing.go:1196 +0x21a
> 
> Previous write at 0x00c000103a21 by goroutine 68:
>   testing.(*common).FailNow()
>       /usr/local/go/src/testing/testing.go:740 +0x4f
>   testing.(*common).Fatalf()
>       /usr/local/go/src/testing/testing.go:815 +0x9d
>   testing.(*T).Fatalf()
>       <autogenerated>:1 +0x75
>   github.com/golang/mock/gomock.(*Controller).Call.func1()
>       /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/vendor/github.com/golang/mock/gomock/controller.go:231 +0x731
>   github.com/golang/mock/gomock.(*Controller).Call()
>       /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/vendor/github.com/golang/mock/gomock/controller.go:247 +0xdc
>   github.com/redhat-developer/odo/pkg/kclient.(*MockClientInterface).ExecCMDInContainer()
>       /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/pkg/kclient/mock_Client.go:315 +0x369
>   github.com/redhat-developer/odo/pkg/remotecmd.ExecuteCommand()
>       /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/pkg/remotecmd/remotecmd.go:36 +0x701
>   github.com/redhat-developer/odo/pkg/remotecmd.(*kubeExecProcessHandler).StartProcessForCommand.func1()
>       /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/pkg/remotecmd/kubeexec.go:101 +0xdb
> 
> Goroutine 67 (running) created at:
>   testing.(*T).Run()
>       /usr/local/go/src/testing/testing.go:1238 +0x5d7
>   github.com/redhat-developer/odo/pkg/remotecmd.TestKubeExecProcessHandler_StartProcessForCommand()
>       /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/pkg/remotecmd/kubexec_test.go:196 +0x6bc
>   testing.tRunner()
>       /usr/local/go/src/testing/testing.go:1193 +0x202
> 
> Goroutine 68 (finished) created at:
>   github.com/redhat-developer/odo/pkg/remotecmd.(*kubeExecProcessHandler).StartProcessForCommand()
>       /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/pkg/remotecmd/kubeexec.go:96 +0x73c
>   github.com/redhat-developer/odo/pkg/remotecmd.TestKubeExecProcessHandler_StartProcessForCommand.func3()
>       /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/pkg/remotecmd/kubexec_test.go:208 +0x3f1
>   testing.tRunner()
>       /usr/local/go/src/testing/testing.go:1193 +0x202
> ==================
> --- FAIL: TestKubeExecProcessHandler_StartProcessForCommand (10.00s)
>     --- FAIL: TestKubeExecProcessHandler_StartProcessForCommand/command_with_all_fields_returned_no_error (10.00s)
>         remotecmd.go:36: Unexpected call to *kclient.MockClientInterface.ExecCMDInContainer([my-container my-pod [/bin/sh -c echo $$ > /opt/odo/.odo_devfile_cmd_my-exec-cmd.pid && cd /path/to/working/dir && export ENV_VAR2='value2' ENV_VAR1='value1' && (tail -f /path/to/a/file) 1>>/proc/1/fd/1 2>>/proc/1/fd/2] 0xc00047a030 0xc00047a040 <nil> false]) at /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/pkg/remotecmd/remotecmd.go:36 because: 
>             expected call at /workspace/dcb6538f-a4d0-4109-85f2-86d8f7960de8/pkg/remotecmd/kubexec_test.go:185 doesn't match the argument at index 2.
>             Got: [/bin/sh -c echo $$ > /opt/odo/.odo_devfile_cmd_my-exec-cmd.pid && cd /path/to/working/dir && export ENV_VAR2='value2' ENV_VAR1='value1' && (tail -f /path/to/a/file) 1>>/proc/1/fd/1 2>>/proc/1/fd/2]
>             Want: is equal to [/bin/sh -c echo $$ > /opt/odo/.odo_devfile_cmd_my-exec-cmd.pid && cd /path/to/working/dir && export ENV_VAR1='value1' ENV_VAR2='value2' && (tail -f /path/to/a/file) 1>>/proc/1/fd/1 2>>/proc/1/fd/2]
>
Whoops - I can see the order of the env vars in the "export ..." part of the command line is different.. This is something that can indeed happen (I changed the handling of the env vars from a list to a map in #5768, which iteration order is not guaranteed).. I can take a look at fixing that in a separate PR if you want.

File: pkg/odo/cli/ui/table.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+			Header: text.FormatUpper,
+			Row:    text.FormatDefault,
+		},
+		Options: table.Options{
+			DrawBorder:      false,
+			SeparateColumns: false,
+			SeparateFooter:  false,
+			SeparateHeader:  false,
+			SeparateRows:    false,
+		},
```
Comment (valaparthvi) [inline pkg/odo/cli/ui/table.go:33]: Optional: This configuration is default, and I think it's safe to drop it.

**Review by valaparthvi — APPROVED** [review]

**Pull request merged**